qiskit/pulse/instructions/frequency.py

```diff
@@ -37,14 +38,14 @@ class SetFrequency(Instruction):
     def __init__(
         self,
         frequency: Union[float, ParameterExpression],
-        channel: PulseChannel,
+        channel: Union[PulseChannel, Frame],
```

Comment:
What do you think about internally converting channel into frame? If this handles two types of representations, we might need to manage many branching logics in the transformer and I'm concerned about inducing some edge cases. This also helps future deprecation of frame operation on channels.

Strictly speaking, frame operation on channels and frames are different due to existence of control channels, but I don't think there is practical trouble with this, i.e. is there any situation we want to decouple target qubit frame from cross resonance?

Of course we can add warning when converting a channel into frame.

Reply:
Perhaps we attach to each `PulseChannel` a `Frame`? This would make it explicit.

Reply:
Perhaps a better approach would be to forbid shift/set frequency/phase on PulseChannels and only allow such instructions to be done through frames. This would avoid conflicts where one tries to set the frequency of the frame and channel to different value simultaneously. Thoughts?

Reply:
I don't see that as a conflict -- I can say `SetFrequency(freq, chan0, t0); SetFrequency(freq2, chan0, t0)` and only the second instruction will have an effect. In the other thread I mentioned a worry about a conflict between one instruction's waveform's frame's channel and its (the instruction's) channel (`Play(signal(wf, frame), chan)`)

Reply:
Typically, a logical element of device operation has a single frame and channel pair (qubit drive, CR drive, 1->2 drive on a qutrit). It feels redundant to specify both frame and channel for every play instruction, so being able to attach a frame to a channel seems nice. One could create a CRChannel, Drive12Channel, etc. However, then one has the burden of defining and caring about these extra channel types.

When would frame instructions on channels be deprecated? When the backend representation supports a concept of frames? Right now, the frame instructions get resolved to instructions on the channels, so it doesn't seem right to give a warning at the moment. Maybe frame operations on channels could be deprecated in the builder?

Reply:
Frame added to PulseChannel in 5ec2682

Reply:
This is more in-line with what @nkanazawa1989 was suggesting: df7ac4e (it's not fully functional yet). I think it feels better to only work with frames when dealing with shift and set phase/frequency as these instructions feel more like frame instructions (a channel can still be given but a warning is raised and the channel is converted to a frame internally). Since each pulse channel now has a frame associated to it we can treat this frame as the native frame of the channel, i.e. any frame instructions (set/shift freq/phase) on this frame will act on the pulse channel on the hardware. Therefore, resolving frames amounts to adding the right set/shift freq/phase on the channel native frames.